test/common/router/scoped_config_impl_test.cc

```diff
@@ -317,6 +345,191 @@ TEST(ScopeKeyBuilderImplTest, Parse) {
+  EXPECT_EQ(info_->routeConfig().get(), route_config_.get());
+  EXPECT_TRUE(TestUtility::protoEqual(info_->configProto(), config_copy));
+  EXPECT_EQ(info_->scopeName(), "foo_scope");
+  EXPECT_EQ(info_->scopeKey(), makeKey({"foo", "bar"}));
+}
+
+class ScopedRouteInfoDeathTest : public ScopedRouteInfoTest {};
+
+TEST_F(ScopedRouteInfoDeathTest, AssertFailure) {
+  EXPECT_DEBUG_DEATH(ScopedRouteInfo(std::move(scoped_route_config_), nullptr),
+                     "ScopedRouteInfo expects a valid RouteConfigProvider.");
+
+  route_config_->name_ = "bar_route";
+  EXPECT_DEBUG_DEATH(
+      ScopedRouteInfo(std::move(scoped_route_config_), std::move(route_config_provider_)),
+      "RouteConfigProvider's name 'bar_route' doesn't match route_configuration_name 'foo_route'.");
+}
+
+class ScopedConfigImplTest : public testing::Test {
+public:
+  void SetUp() override {
+    std::string yaml_plain = R"EOF(
+  fragments:
+  - header_value_extractor:
+      name: 'foo_header'
+      element_separator: ','
+      element:
+        key: 'bar'
+        separator: '='
+  - header_value_extractor:
+      name: 'bar_header'
+      element_separator: ';'
+      index: 2
+)EOF";
+    TestUtility::loadFromYaml(yaml_plain, key_builder_config_);
+
+    scope_info_a_ = makeScopedRouteInfo(R"EOF(
+    name: foo_scope
+    route_configuration_name: foo_route
+    key:
+      fragments:
+        - string_key: foo
+        - string_key: bar
+)EOF");
+    scope_info_a_v2_ = makeScopedRouteInfo(R"EOF(
+    name: foo_scope
+    route_configuration_name: foo_route
+    key:
+      fragments:
+        - string_key: xyz
+        - string_key: xyz
+)EOF");
+    scope_info_b_ = makeScopedRouteInfo(R"EOF(
+    name: bar_scope
+    route_configuration_name: bar_route
+    key:
+      fragments:
+        - string_key: bar
+        - string_key: baz
+)EOF");
+  }
+  std::shared_ptr<ScopedRouteInfo> makeScopedRouteInfo(const std::string& route_config_yaml) {
+    envoy::api::v2::ScopedRouteConfiguration scoped_route_config;
+    TestUtility::loadFromYaml(route_config_yaml, scoped_route_config);
+
+    std::shared_ptr<MockConfig> route_config = std::make_shared<NiceMock<MockConfig>>();
+    route_config->name_ = scoped_route_config.route_configuration_name();
+
+    std::unique_ptr<MockRouteConfigProvider> route_config_provider =
+        std::make_unique<NiceMock<MockRouteConfigProvider>>();
+    EXPECT_CALL(*route_config_provider, config()).WillRepeatedly(Return(route_config));
+    return std::make_shared<ScopedRouteInfo>(std::move(scoped_route_config),
+                                             std::move(route_config_provider));
+  }
+
+  std::shared_ptr<ScopedRouteInfo> scope_info_a_;
+  std::shared_ptr<ScopedRouteInfo> scope_info_a_v2_;
+  std::shared_ptr<ScopedRouteInfo> scope_info_b_;
+  ScopedRoutes::ScopeKeyBuilder key_builder_config_;
+  std::unique_ptr<ScopedConfigImpl> scoped_config_impl_;
+};
+
+TEST_F(ScopedConfigImplTest, PickRoute) {
```

Comment:
Please `//` single line comment above each test explaining what it is about.

Reply:
done